src/Agent.Worker/TestResults/ResultsCommands.cs

```diff
@@ -0,0 +1,321 @@
+                buildFlavor: runContext != null ? runContext.Configuration : string.Empty,
+                buildPlatform: runContext != null ? runContext.Platform : string.Empty
+                );
+
+            testRunData.Attachments = runAttachments.ToArray();
+
+            //publish run if there are results.
+            if (runResults.Count > 0)
+            {
+                publisher.StartTestRun(testRunData);
+                publisher.AddResults(runResults.ToArray());
+                String webAccessUrl = publisher.EndTestRun(publishAttachmentsAsArchive: true);
+                executionContext.Debug(String.Format(CultureInfo.CurrentCulture, "Published Test Run : {0}", webAccessUrl));
+            }
+        }
+
+        /// <summary>
+        /// Publish separate test run for each result file that has results.
+        /// </summary>
+        private void PublishToNewTestRunPerTestResultFile(List<string> resultFiles, TestRunPublisher publisher)
+        {
+            int runCount = 1;
+            // Publish separate test run for each result file that has results.
+            foreach (string resultFile in resultFiles)
+            {
+                string runName = null;
+                if (!String.IsNullOrWhiteSpace(runTitle))
+                {
+                    runName = resultFiles.Count > 1 ? String.Format(CultureInfo.InvariantCulture, "{0} {1}", runTitle, runCount++) : runTitle;
+                }
+
+                TestRunData testRunData = publisher.ReadResultsFromFile(resultFile, runName);
+
+                if (testRunData != null && testRunData.Results != null && testRunData.Results.Length > 0)
+                {
+                    publisher.StartTestRun(testRunData);
+                    publisher.AddResults(testRunData.Results);
+                    String webAccessUrl = publisher.EndTestRun();
+                    executionContext.Debug(String.Format(CultureInfo.CurrentCulture, "Published Test Run : {0}", webAccessUrl));
+                }
+                else
+                {
+                    var message = StringUtil.Loc("InvalidResultFiles", resultFile);
+                    warnings.Add(message);
+                    executionContext.Warning(message);
+                }
+            }
+        }
+
+        private IResultReader GetTestResultReader(string testRunner)
+        {
+            IResultReader reader;
+            if (testRunner.Equals("VSTest", StringComparison.OrdinalIgnoreCase))
+            {
+                reader = new TrxResultReader(executionContext);
+            }
+            else if (testRunner.Equals("JUnit", StringComparison.OrdinalIgnoreCase))
+            {
+                reader = new JUnitResultReader(executionContext);
+            }
+            else if (testRunner.Equals("NUnit", StringComparison.OrdinalIgnoreCase))
+            {
+                reader = new NUnitResultReader(executionContext);
+            }
```

Comment:
You can use IExtensionManager to avoid this.
You could simply call in oneliner

`extensionManager.GetExtension<IResultReader>()).FirstOrDefault(x => x.Type.Equals(testRunner);`

Take a look at the ReleaseJobExtension, we have multiple implementation of IArtifactProvider and pick the right implementation by the type and avoid this if-else-if blocks.